tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,324 @@
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!("acquired semaphore after value was already initialized.");
+                    }
+                }
+                Err(_) => {
+                    if self.initialized() {
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!(
+                            "Semaphore closed, but the OnceCell has not been initialized."
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Since `get` gives us access to immutable references of the
+// OnceCell, OnceCell can only be Sync if T is Sync, otherwise
+// OnceCell would allow sharing references of !Sync values across
+// threads. We need T to be Send in order for OnceCell to by Sync
+// because we can use `set` on `&OnceCell<T>` to send
+// values (of type T) across threads.
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+
+// Access to OnceCell's value is guarded by the semaphore permit
+// and atomic operations on `value_set`, so as long as T itself is Send
+// it's safe to send it to another thread
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+/// Error returned from the [`OnceCell::set`] method
+///
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub struct AlreadyInitializedError;
+
+impl fmt::Display for AlreadyInitializedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "AlreadyInitializedError")
+    }
+}
+
+impl Error for AlreadyInitializedError {}
+
+/// Error returned from the [`OnceCell::get`] method
+///
+/// [`OnceCell::get`]: crate::sync::OnceCell::get
+#[derive(Debug, PartialEq)]
+pub struct NotInitializedError;
+
+impl fmt::Display for NotInitializedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "NotInitializedError")
+    }
+}
+
+impl Error for NotInitializedError {}
+
+/// Allows one to lazily call an async function
+///
+/// # Examples
+/// ```
+/// use tokio::sync::Lazy;
+///
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// static LAZY : Lazy<u32> = Lazy::new(|| Box::pin(some_computation()));
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let result = tokio::spawn(async {
+///         LAZY.get().await
+///     }).await.unwrap();
+///     assert_eq!(*result, 2);
+/// }
+/// ```
+pub struct Lazy<T, F = fn() -> Pin<Box<dyn Future<Output = T> + Send>>> {
+    cell: OnceCell<T>,
+    init: UnsafeCell<Option<F>>,
```

Comment:
I don't understand why this is not just
```suggestion
    init: F,
```

Reply:
If `F` was `FnOnce`, it would be necessary. Is there a reason why `F` must be called more than once?

Reply:
Related: Is it correct that currently, if one future is dropped (one task isn't interested in the result anymore), the next computation will start again from the beginning?

Reply:
It must be `Fn` in case the operation is cancelled. And to the other question, yes as well.

Reply:
And there's no other way? Something like storing the future inside of the Lazy, so that it can be continued even if the first caller is dropped?

Reply:
Well maybe .. but you would need a full unsafe implementation for it similar to `OnceCell`. And `OnceCell` would not be able to have the same feature.

Reply:
Unless we modify the implementation of OnceCell as well: 
* Store a `Future<Output = Result<_,_>>` inside the `OnceCell`.
* Insert the `Future` of `get_or_init` only if there's no `Future` present.
* The first waiting task tries to drive the future to completion.
* If the task is dropped, the next one continues the work on the same `Future`.
* If the `Future` completes with an `Err`, it is removed from the `OnceCell`, and the next available `Future` will be inserted.
* If the `Future` completes with an `Ok`, we are ready.
* There is a function to join the waiting for the result without having to provide an init fn.

But this would be yet harder to implement (if possible at all?)